File: typescript/src/application/handlers/UserRegistrationHandler.ts
import type { Handler } from '../Handler'
import type { UserRepository } from '../repositories/UserRepository'
import type { UsernameService } from '../../domain/services/UsernameService'
import type { UuidWrapper } from '../wrapper/UuidWrapper'
import type { CreateUser } from '../../domain/commands/CreateUser'
import { User } from '../../domain/aggregates/User'

type UserId = string

export class UserRegistrationHandler implements Handler<CreateUser, UserCreated> {
  constructor(
    private usernameService: UsernameService,
    private userRepository: UserRepository,
    private uuidWrapper: UuidWrapper
  ) {}

  handle = async (createUser: CreateUser) => {
    if (await this.usernameService.usernameExists(createUser.username)) {
      throw new UsernameAlreadyInUseError()
    }

    const user = User.create(this.uuidWrapper.generateUuid(), createUser)
    await this.userRepository.save(user)

    return {
      id: user.userId,
      username: createUser.username,
      about: createUser.about,
    }
  }
}

export type UserCreated = {
  id: UserId
  username: string
  about: string
}

export class UsernameAlreadyInUseError extends Error {}

File: scala/src/main/scala/openchat/application/handlers/UserRegistrationHandler.scala
package openchat.application.handlers

import openchat.application.Handler
import openchat.application.repositories.UserRepository
import openchat.application.wrapper.UuidWrapper
import openchat.domain.aggregates.User
import openchat.domain.commands.CreateUser
import openchat.domain.errors.{UserRegistrationError, UsernameAlreadyExists}
import openchat.domain.services.UsernameService

import scala.concurrent.{ExecutionContext, Future}

final class UserRegistrationHandler(
  usernameService: UsernameService,
  userRepository: UserRepository,
  uuidWrapper: UuidWrapper
)(implicit ec: ExecutionContext)
  extends Handler[CreateUser, Either[UserRegistrationError, UserCreated]] {

  def handle(createUser: CreateUser): Future[Either[UserRegistrationError, UserCreated]] = {
    for {
      exists <- usernameService.usernameExists(createUser.username)
      result <- if (exists) {
        Future.successful(Left(UsernameAlreadyExists))
      } else {
        val user = User.create(uuidWrapper.generateUuid(), createUser)
        userRepository.save(user).map(_ =>
          Right(UserCreated(id = user.userId, username = createUser.username, about = createUser.about))
        )
      }
    } yield result
  }
}

final case class UserCreated(id: String, username: String, about: String)
class UsernameAlreadyInUseError extends RuntimeException
